Source/santad/Logs/SNTSimpleMaildir.m

```diff
@@ -0,0 +1,348 @@
+  [self flush];
+}
+
+/** Fires the flush timer programmatically. Used for testing purposes. */
+- (void)fireFlushTimer {
+  [_flushTimer fire];
+}
+
+/**
+ * Flushes out buffered data.
+ *
+ * Returns the number of events that we attempted to flush, and populates the error if that flush
+ * failed.
+ */
+- (void)flushLockedWithError:(NSError **)error {
+  NSAssert([_outputProto.recordsArray count] < INT_MAX, @"Too many records");
+
+  if ([_outputProto.recordsArray count] == 0) {
+    return;
+  }
+
+  if (![self createSpoolDirectoriesWithError:error]) {
+    return;
+  }
+
+  if (_estimatedSpoolSize > _spoolSizeThreshold) {
+    _estimatedSpoolSize = [self spoolDirectorySizeWithError:error];
+    if (_estimatedSpoolSize > _spoolSizeThreshold) {
+      if (error) {
+        *error = MakeError(@"file_system_threshold_exceeded");
+      }
+      return;
+    }
+  }
+
+  NSString *filename = [NSString stringWithFormat:@"%@.%u", _filenamePrefix, _createdFileCount];
+  NSString *exposedFilepath = [_newDirectory stringByAppendingPathComponent:filename];
+  NSString *outputFilepath = [_tmpDirectory stringByAppendingPathComponent:filename];
+  NSOutputStream *outputFile = [NSOutputStream outputStreamToFileAtPath:outputFilepath append:NO];
+  [outputFile open];
+  _createdFileCount++;
+
+  if (!outputFile) {
+    if (error) {
+      *error = MakeError(@"data_loss_on_open");
+    }
+    return;
+  }
+
+  BOOL writeSuccess = NO;
+  @try {
+    [_outputProto writeToOutputStream:outputFile];
+    writeSuccess = YES;
+  } @catch (NSException *exception) {
+    NSLog(@"Error while writing to %@: %@", outputFilepath, exception);
+    if (error) {
+      *error = MakeError(@"data_loss_on_write");
+    }
+  }
+
+  [outputFile close];
+
+  if (!writeSuccess) {
+    // Unable to successfully write all data.
+    [[NSFileManager defaultManager] removeItemAtPath:outputFilepath error:nil];
+    return;
+  }
+
+  if (![[NSFileManager defaultManager] moveItemAtPath:outputFilepath
+                                               toPath:exposedFilepath
+                                                error:nil]) {
+    // Delete the tmp file if unable to move
+    [[NSFileManager defaultManager] removeItemAtPath:outputFilepath error:nil];
+    if (error) {
+      *error = MakeError(@"data_loss_on_move");
+    }
+  }
+
+  if (error && !*error) {
+    _estimatedSpoolSize += _outputProtoSerializedSize;
+  }
+
+  return;
+}
+
+- (void)flushAndUpdateCountersLocked {
+  NSError *error = nil;
+  [self flushLockedWithError:&error];
+  [_eventsFlushedCounter incrementBy:[_outputProto.recordsArray count]
+                      forFieldValues:@[ ErrorToMetricFieldName(error) ]];
+
+  // Clear output buffer.
+  _outputProto = [[SNTPBLogBatch alloc] init];
+  _outputProtoSerializedSize = 0;
+}
+
+- (void)flush {
+  dispatch_sync(_flushQueue, ^{
+    [self flushAndUpdateCountersLocked];
+  });
+}
+
+- (BOOL)createDirectory:(NSString *)dir withError:(NSError **)error {
```

Comment:
With that said... This code I ported over should probably set `withIntermediateDirectories:YES` instead. Otherwise it would be difficult to configure arbitrary paths

Reply:
Adopted this change after moving to `withIntermediateDirectories:YES`

Reply:
Hm, going back and forth here - reverted back to the original to save the need to do a path check twice in the event the directory already exists. Instead added comments to hopefully clarify the multiple paths through the code.